PORTING.md

```diff
@@ -0,0 +1,351 @@
+```ini
+[openocd]
+configfile = board/atmel_sam4l8_xplained_pro.cfg
+```
+For custom setups, [read this](http://xpcc.io/reference/build-system/#project-configuration).
+
+You should now be able to program the device using `scons program`.
+Check if the LED is blinking at a constant interval.
+
+If the LED is not blinking and you have OpenOCD connected, you can [launch a GDB debug session](https://github.com/roboterclubaachen/xpcc/tree/develop/examples/stm32f3_discovery/gdb).
+
+If something goes wrong in the building step, check the previous steps, maybe you missed something, or we forgot to add something to this guide.
+If you're totally stuck write us an email (try to work the problem first though ;-).
+
+### Porting the delay functions
+
+To keep it simple, we'll leave the device running on the speed it boot up on. This is usually some internal oscillator in the MHz range.
+In order to give us an accurate busy-wait delay functions (`xpcc::delay{Nano|Micro|Milli}seconds(uint16_t)`) that we can use for blinking an LED and bit-banging GPIO, the clock driver needs to initialize four values:
+
+- `fcpu` contains the CPU frequency in Hz
+- `fcpu_kHz` contains the CPU frequency in kHz
+- `fcpu_MHz` contains the CPU frequency in MHz
+- `ns_per_loop` contains the ns per busy-wait delay loop (processor dependent)
+
+Create two files in the `xpcc/architecture/platform/driver/clock/sam4` folder:
+
+- `clock.hpp` (see [STM32 version for verbatim](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/clock/stm32/clock.hpp.in?ts=4)):
+```cpp
+// your copyright with bsd license
+#include <stdint.h>
+#include "../../../device.hpp"
+#include "../generic/common_clock.hpp"
+
+using namespace xpcc::clock;
+```
+
+- `clock.cpp` ([STM32 version](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/clock/stm32/clock.cpp.in?ts=4)) (assuming 8 MHz initial clock):
+```cpp
+// your copyright with bsd license
+#include "clock.hpp"
+
+namespace xpcc
+{
+namespace clock
+{
+
+uint32_t ATTRIBUTE_FASTDATA fcpu(8000000);
+uint32_t ATTRIBUTE_FASTDATA fcpu_kHz(8000);
+uint16_t ATTRIBUTE_FASTDATA fcpu_MHz(8);
+// Cortex-M0: 4000 loops per MHz
+// Cortex-M7: 1000 loops per MHz
+// otherwise: 3000 loops per MHz
+uint16_t ATTRIBUTE_FASTDATA ns_per_loop asm("ns_per_loop") (3000 / 8);
+
+}
+}
+```
+
+If you [look at the STM32 `driver.xml` file](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/clock/stm32/driver.xml#L19-L22), you'll see that the internal high- and low-speed oscillator values differ even within the same platform. These values are therefore initialized in the clock driver and not the core driver.
+
+### Porting the GPIO driver
+
+> In xpcc, GPIOs are assumed to be usable without explicitly initializing the subsystem.
+> For other peripheral IP, the clock is enabled in the `initialize` method that needs to be explicitly called by the user.
```

Comment:
Ah, I thought it was port of our documentation and therefore quoted. Will inline then.